### ISSUE TITLE
Express Error Handling Overwriting Other Patch

### ISSUE THREAD

**Issue by emhagman** [DataDog/dd-trace-js#548]
I use the package https://www.npmjs.com/package/express-async-errors to use `async` handlers in my code. However, Datadog is patching the same `Layer.prototype.handle` method as this library. When my error is thrown, both are caught but Datadog is killing the connection and causing "Network Errors" to occur instead of my middleware handling the error properly.

Instead of returning with a 500, I get a SPDY_PROTOCOL_ERROR inside of Chrome.

Any ideas?

**Comment by rochdev** [comment]
I don't know this library. I'll have to take a look where the conflict happens. Does this happen for any error thrown from any middleware?

**Comment by emhagman** [comment]
I can write a test case to see. I can reproduce it locally when `dd-trace-js` is enabled every time. The weird thing is my middleware can still catch the error but without fail, I get a SPDY_PROTOCOL_ERROR as if the connection is just killed. Turning it off makes everything work. I'll get back to you on the any middleware thing.

**Comment by emhagman** [comment]
It only occurs when using `async`, even if the `express-async-errors` module enabled.

```
TypeError: Cannot read property 'id' of undefined
drafted-api      |     at Layer.<anonymous> (/usr/src/app/src/routes/matches.js:668:25)
drafted-api      |     at Layer.newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at web.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:115:21)
drafted-api      |     at Scope._activate (/usr/src/app/node_modules/dd-trace/src/scope/new/scope.js:38:14)
drafted-api      |     at Scope.activate (/usr/src/app/node_modules/dd-trace/src/scope/new/base.js:13:17)
drafted-api      |     at Object.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/util/web.js:90:27)
drafted-api      |     at callHandle (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:113:14)
drafted-api      |     at /usr/src/app/node_modules/dd-trace/src/plugins/router.js:62:14
drafted-api      |     at newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at Layer.handle [as handle_request] (/usr/src/app/node_modules/express/lib/router/layer.js:95:5)
drafted-api      |     at next (/usr/src/app/node_modules/express/lib/router/route.js:137:13)
drafted-api      |     at Route.dispatch (/usr/src/app/node_modules/express/lib/router/route.js:112:3)
drafted-api      |     at Layer.newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at web.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:115:21)
drafted-api      |     at Scope._activate (/usr/src/app/node_modules/dd-trace/src/scope/new/scope.js:38:14)
drafted-api      |     at Scope.activate (/usr/src/app/node_modules/dd-trace/src/scope/new/base.js:13:17)
drafted-api      |     at Object.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/util/web.js:90:27)
drafted-api      |     at callHandle (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:113:14)
drafted-api      |     at /usr/src/app/node_modules/dd-trace/src/plugins/router.js:62:14
drafted-api      |     at newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at Layer.handle [as handle_request] (/usr/src/app/node_modules/express/lib/router/layer.js:95:5)
drafted-api      |     at /usr/src/app/node_modules/express/lib/router/index.js:281:22
drafted-api      |     at Function.process_params (/usr/src/app/node_modules/express/lib/router/index.js:335:12)
drafted-api      |     at Function.processParamsWithTrace [as process_params] (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:35:28)
drafted-api      |     at next (/usr/src/app/node_modules/express/lib/router/index.js:275:10)
drafted-api      |     at /usr/src/app/node_modules/dd-trace/src/plugins/router.js:98:18
drafted-api      |     at Layer.serveStatic (/usr/src/app/node_modules/serve-static/index.js:75:16)
drafted-api      |     at Layer.newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at web.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:115:21)
drafted-api      |     at Scope._activate (/usr/src/app/node_modules/dd-trace/src/scope/new/scope.js:38:14)
drafted-api      |     at Scope.activate (/usr/src/app/node_modules/dd-trace/src/scope/new/base.js:13:17)
drafted-api      |     at Object.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/util/web.js:90:27)
drafted-api      |     at callHandle (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:113:14)
drafted-api      |     at /usr/src/app/node_modules/dd-trace/src/plugins/router.js:62:14
drafted-api      |     at newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at Layer.handle [as handle_request] (/usr/src/app/node_modules/express/lib/router/layer.js:95:5)
drafted-api      |     at trim_prefix (/usr/src/app/node_modules/express/lib/router/index.js:317:13)
drafted-api      |     at /usr/src/app/node_modules/express/lib/router/index.js:284:7
drafted-api      |     at Function.process_params (/usr/src/app/node_modules/express/lib/router/index.js:335:12)
drafted-api      |     at Function.processParamsWithTrace [as process_params] (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:35:28)
drafted-api      |     at next (/usr/src/app/node_modules/express/lib/router/index.js:275:10)
drafted-api      |     at /usr/src/app/node_modules/dd-trace/src/plugins/router.js:98:18
drafted-api      |     at Layer.serveStatic (/usr/src/app/node_modules/serve-static/index.js:75:16)
drafted-api      |     at Layer.newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at web.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:115:21)
drafted-api      |     at Scope._activate (/usr/src/app/node_modules/dd-trace/src/scope/new/scope.js:38:14)
drafted-api      |     at Scope.activate (/usr/src/app/node_modules/dd-trace/src/scope/new/base.js:13:17)
drafted-api      |     at Object.wrapMiddleware (/usr/src/app/node_modules/dd-trace/src/plugins/util/web.js:90:27)
drafted-api      |     at callHandle (/usr/src/app/node_modules/dd-trace/src/plugins/router.js:113:14)
drafted-api      |     at /usr/src/app/node_modules/dd-trace/src/plugins/router.js:62:14
drafted-api      |     at newFn (/usr/src/app/node_modules/express-async-errors/index.js:16:20)
drafted-api      |     at Layer.handle [as handle_request] (/usr/src/app/node_modules/express/lib/router/layer.js:95:5)
```

**Comment by rochdev** [comment]
Thank you for sharing these, it will be helpful. I think there must be some assumption that a middleware always calls a callback, and since the library changes this behavior, our instrumentation doesn't handle it properly.

**Comment by emhagman** [comment]
Thanks for looking into it! I also tried reversing the order, and having `express-async-errors` patch first and then `dd-trace` but the same thing still occurs. For now I have disabled `dd-trace`

**Comment by emhagman** [comment]
It seems like the error is being thrown twice. If I add another middleware that comes after my **errors** middleware, I discard the error if I handled it.

See below for the catch-all. It seems really weird I need to catch the error again in another middleware and again this still only happens with Datadog

```
module.exports = function errorCatchAll(err: Object, req: Object, res: Object, next: Function) {
    if (!req._errorHandled) {
        Logger.error(`Uncaught error was caught in the catch-all`);
        return next(err);
    } else {
        // Don't call next middleware, we handled it
        Logger.debug(`Handled error made it past the "errors" middleware, catching it`);
    }
};
```

**Comment by rochdev** [comment]
I took a quick look yesterday and was able to easily reproduce it. I'll have to investigate further to see how we can handle this. I'll try to at least isolate it in an actual unit test.

**Comment by emhagman** [comment]
Thanks, I appreciate the update 👍

**Comment by emhagman** [comment]
To update my previous comment, it appears to fix things in that, my **non-20x** responses don't break the connection; however, those errors are not showing up in Datadog APM anymore, probably because I am not letting them reach the Datadog error middleware/hooks.

**Comment by rochdev** [comment]
That is a totally different issue however. Do you expect errors caught by `express-async-errors` to appear in Datadog? If that's the case we probably need to instrument the library. The current implementation only catches uncaught exceptions and assumes that if it was caught it means it was handled regardless of the status code. It's still possible to add error metadata to the request span manually however. I can show you how to do that once the main issue is fixed.

**Comment by emhagman** [comment]
Nevermind, a 500 error was just traced and the error was caught so our code must've still been deploying.

I guess the main point here is that errors "caught" by `express-async-errors` are supposed to be same as if I threw an error in a non-async request call back. It is not a middleware that is supposed to catch errors and swallow them, it simply passes them along to the next ExpressJS error middleware, as if it were uncaught.

**Comment by rochdev** [comment]
Yes sorry you are correct. I was thinking of global uncaught exceptions which this library is basically handling. Then it should work when fixed on our end.

Just to confirm, your hack to add a middleware works _exactly_ like if `dd-trace` is disabled?

**Comment by emhagman** [comment]
Correct, working and tracing is enabled and sending data. I believe the error is thrown 2x so I added the middleware above to catch it again and ignore it if I marked it as handled in my original error middleware.

Only difference I can see is with `dd-trace` disabled the error isn't thrown 2x. I say it's being thrown twice because the error is caught in my middleware and printed through our logger (pinojs) and then white text of the stack appears again outside of my caught logging of the error, which appears just be standard stdout. When disabled there is no white stack trace.

**Comment by rochdev** [comment]
I opened a PR that should fix the issue, but I wasn't able to setup a unit test for it.

@emhagman Can you try from the branch and let me know if it fixes all the issues mentioned above?

**Comment by rochdev** [comment]
The fix has now been released in 0.11.1

**Comment by emhagman** [comment]
Just saw all this, I'll give it a try, thanks!
